distogram/__init__.py

```diff
@@ -1,182 +1,152 @@
+    m = (ci * (v - c) ** n for i, (ci, v) in enumerate(zip(counts, x)))
     return sum(m) / sum(counts)
 
 
-def _weighted_diff(h, l, r):
-    diff = l[0] - r[0]
+def _weighted_diff(h: Distogram, left: Bin, right: Bin):
+    diff = left[0] - right[0]
     if h.weighted_diff is True:
-        diff *= math.log(0.00001 + min(l[1], r[1]))
+        diff *= math.log(EPSILON + min(left[1], right[1]))
     return diff
 
 
-def _update_diffs(h, i):
+def _update_diffs(h: Distogram, i: int) -> None:
     if h.diffs is not None:
         update_min = False
+
         if i > 0:
-            diff = _weighted_diff(h, h.bins[i], h.bins[i-1])
-            if h.diffs[i-1] == h.min_diff:
-                h.diffs[i-1] = diff
+            if h.diffs[i - 1] == h.min_diff:
                 update_min = True
-            else:
-                h.diffs[i-1] = diff
-                if h.diffs[i-1] < h.min_diff:
-                    h.min_diff = h.diffs[i-1]
+
+            h.diffs[i - 1] = _weighted_diff(h, h.bins[i], h.bins[i - 1])
+            if h.diffs[i - 1] < h.min_diff:
+                h.min_diff = h.diffs[i - 1]
+
         if i < len(h.bins) - 1:
-            diff = _weighted_diff(h, h.bins[i+1], h.bins[i])
             if h.diffs[i] == h.min_diff:
-                h.diffs[i] = diff
                 update_min = True
-            else:
-                h.diffs[i] = diff
-                if h.diffs[i] < h.min_diff:
-                    h.min_diff = h.diffs[i]
+
+            h.diffs[i] = _weighted_diff(h, h.bins[i + 1], h.bins[i])
+            if h.diffs[i] < h.min_diff:
+                h.min_diff = h.diffs[i]
 
         if update_min is True:
             h.min_diff = min(h.diffs)
 
+    return
+
 
-def _trim(h, index):
-    bins = h.bins
-    while len(bins) > h.bin_count:
+def _trim(h: Distogram) -> Distogram:
+    while len(h.bins) > h.bin_count:
         if h.diffs is not None:
-            min_diff = h.min_diff
-            i = h.diffs.index(min_diff)
+            i = h.diffs.index(h.min_diff)
         else:
-            min_diff = None
-            i = None
-            for index, value in enumerate(h.bins):
-                if index > 0:
-                    diff = _weighted_diff(h, h.bins[index], h.bins[index-1])
-                    if min_diff is None or diff < min_diff:
-                        min_diff = diff
-                        i = index - 1
-
-        bins[i] = (
-            (bins[i][0]*bins[i][1] + bins[i+1][0]*bins[i+1][1])
-            / (bins[i][1] + bins[i+1][1]),
-            bins[i][1] + bins[i+1][1]
-        )
-        del bins[i+1]
+            diffs = [
+                (i - 1, _weighted_diff(h, b, h.bins[i - 1]))
+                for i, b in enumerate(h.bins[1:], start=1)
+            ]
+            i, _ = min(diffs, key=itemgetter(1))
+
+        v1, f1 = h.bins[i]
+        v2, f2 = h.bins[i + 1]
+        h.bins[i] = (v1 * f1 + v2 * f2) / (f1 + f2), f1 + f2
```

Comment:
There is the potential for a divide-by-zero error here because a user can call `update` with `count=0`. We should check for this either in `update` or here. Which would you prefer?